src/webgpu/api/operation/memory_sync/texture/same_subresource.spec.ts

```diff
@@ -2,21 +2,465 @@ export const description = `
+          usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+        });
+
+        this.queue.writeBuffer(buffer, 0, fullTexelData);
+
+        assert(this.commandEncoder !== undefined);
+        this.commandEncoder.copyBufferToTexture(
+          { buffer, bytesPerRow },
+          { texture: this.texture },
+          this.kTextureSize
+        );
+        break;
+      }
+      case 'attachment-resolve':
+      case 'storage':
+        // [1] Finish implementation
+        throw new SkipTestCase('unimplemented');
+      case 't2b-copy':
+      case 'sample':
+        unreachable();
+    }
+  }
+
+  // Ensure that all encoded commands are finished and subitted.
+  ensureSubmit() {
+    this.ensureContext('queue');
+    this.flushEncodedCommands();
+  }
+
+  private popContext(): GPURenderBundle | GPUCommandBuffer | null {
+    switch (this.currentContext) {
+      case 'queue':
+        unreachable();
+        break;
+      case 'command-encoder': {
+        assert(this.commandEncoder !== undefined);
+        const commandBuffer = this.commandEncoder.finish();
+        this.commandEncoder = undefined;
+        this.currentContext = 'queue';
+        return commandBuffer;
+      }
+      case 'compute-pass-encoder':
+        assert(this.computePassEncoder !== undefined);
+        this.computePassEncoder.endPass();
+        this.computePassEncoder = undefined;
+        this.currentContext = 'command-encoder';
+        break;
+      case 'render-pass-encoder':
+        assert(this.renderPassEncoder !== undefined);
+        this.renderPassEncoder.endPass();
+        this.renderPassEncoder = undefined;
+        this.currentContext = 'command-encoder';
+        break;
+      case 'render-bundle-encoder': {
+        assert(this.renderBundleEncoder !== undefined);
+        const renderBundle = this.renderBundleEncoder.finish();
+        this.renderBundleEncoder = undefined;
+        this.currentContext = 'render-pass-encoder';
+        return renderBundle;
+      }
+    }
+    return null;
+  }
+
+  private makeDummyAttachment(): GPURenderPassColorAttachment {
+    const texture = this.device.createTexture({
+      format: this.kTextureFormat,
+      size: this.kTextureSize,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+    return {
+      view: texture.createView(),
+      loadValue: 'load',
+      storeOp: 'store',
+    };
+  }
+
+  private ensureContext(context: OperationContext) {
+    // Find the common ancestor. So we can transition from currentContext -> context.
+    const ancestorContext =
+      kOperationContexts[
+        Math.min(
+          kOperationContexts.indexOf(context),
+          kOperationContexts.indexOf(this.currentContext)
+        )
+      ];
+
+    // Pop the context until we're at the common ancestor.
+    while (this.currentContext !== ancestorContext) {
+      // We're about to pop another context. Make sure we've
+      // executed any previous render bundles / command buffers.
```

Comment:
Why flush instead of just appending to the list of encodedCommands? Seems like it might be best to keep all of the submits as close together as possible, for the sake of testing synchronization.

Reply:
Fixed; and now it does .push()